### ISSUE TITLE
Improvement ideas

### ISSUE THREAD

**Issue by sactage** [sactage/chatbot-rb#12]
@rinzler135 @ty-a

Any improvements, cleanup, etc. you guys think might be worth doing to the framework?

**Comment by ty-a** [comment]
Gaz Lloyd> basically everything that the irc bots do
Gaz Lloyd> 7:^]
Gaz Lloyd> timestamps in tells would be nice

**Comment by sactage** [comment]
what a coincidence, I just added timestamps earlier today

**Comment by Rinzler135** [comment]
I've made a couple of plugins, but they aren't particularly remarkable and are mainly geared towards the chat the bot resides in. There are 2 problems which I'm confused by though.

Occasionally, the bot will claim that it hasn't !seen someone, but when the command is repeated, with the same syntax, it works fine.

The second one, which is slightly more serious, is when the error `is? is undefined`, which generally requires a reboot.

**Comment by sactage** [comment]
Would you perchance have a full traceback of the second point you brought up?

As for !seen - I don't exactly know what might cause, but try to look into it this weekend.

**Comment by Rinzler135** [comment]
I will attempt to obtain one if possible.

**Comment by Rinzler135** [comment]
An improvement (possibly) - stop `join` from triggering if someone's connection is lost for a few seconds. The bot is currently saying something if they have less than a certain number of edits, and it sometimes says if they haven't visibly left.

Also, it sometimes seems to miss people leaving, and tells people it is still there.

**Comment by sactage** [comment]
Should have fixed the issue with the fatal in e737f47e7

Join/Parts are probably my least favourite thing to deal with for chat. They're unreliable as hell, and there's times where the server will send multiple join events for the same user without any part events, and vice versa... the way the actual Special:Chat deals with this is silly, too. _Maybe_ we can get them more accurate - but it won't be pretty.

**Comment by Rinzler135** [comment]
The only solution I could think of is checking when the last time the user was seen, and if it was <5 seconds ago or so don't count it. I doubt is very elegant/the best solution and I'm not sure what the impact would be.

On a different topic, being able to get the bot to send pm's to people would be really handy, even if it can't receive commands through pm's.

**Comment by sactage** [comment]
Oh - you know what, I was wrong when I said joins/parts are my least favourite thing in chat.

That title should rather go to PMs. PMs require you to connect to an _entirely new chat room_, and poll the server for _every room you have open_ in separate requests. I guess, theoretically, having the bot only handle sending PMs would be less effort - then we wouldn't need to poll the server for anything but the initial room data - but I'm not entirely sure about that, because PMs _might_ require you to ping the server separately from the main room, which could maybe throw a wrench into our plans. That being said - I'll look into it.

**Comment by Rinzler135** [comment]
I think I managed to get as far as opening a new room. I never succeeded in sending any messages through.

**Comment by sactage** [comment]
Yeah, when I tried a while ago I could open new rooms - that's trivial. The annoying bit is interacting with those rooms, since I think it might require a new, unique session id or chat key

**Comment by KORN1UX** [comment]
What about longer sessions for logging? I've tried make a wrapper with daemons gem but it's useless.
Thing is bot keeps connection for a not long time.
